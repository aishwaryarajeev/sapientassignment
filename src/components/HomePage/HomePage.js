import React, { useState } from "react";
import Launches from "../Launches";
import { makeStyles } from "@material-ui/styles";
import * as CONSTANTS from "../../constants";

export const useStyles = makeStyles(() => ({
  mainContainer: {
    height: "100%",
    maxWidth: "1440px",
    backgroundColor: "#e6e6e6",
    margin: "auto",
    "@media (max-width: 700px) ": {
      width: "min-content",
      margin: "25px auto",
    },
  },
  header: {
    fontSize: "20px",
    fontWeight: 600,
    padding: "5px 10px",
  },
  itemContainer: {
    display: "flex",
    justifyContent: "center",
    "@media (max-width: 700px) ": {
      display: "flex",
      flexDirection: "column",
      width: "100%",
      justifyContent: "center",
      alignItems: "center",
    },
  },
  filters: {
    position: "relative",
    width: "200px",
    background: "white",
    height: "550px",
    margin: "5px 10px",
    borderRadius: "5px",
    fontWeight: 600,
    padding: "5px",
    fontSize: "14px",
  },
  launchItems: {
    width: "75%",
    "@media (max-width: 700px) ": {
      width: "220px",
      justifyContent: "center",
      alignItems: "center",
    },
  },
  filterTitle: {
    textAlign: "center",
    fontSize: "12px",
    fontWeight: 400,
    padding: "5px",
    "&:after": {
      content: '" "',
      position: "absolute",
      left: "40px",
      width: "125px",
      height: "18px",
      borderBottom: "1px solid grey",
    },
  },
  filterItems: {
    display: "flex",
    flexWrap: "wrap",
    justifyContent: "center",
  },
  buttonStyle: {
    borderRadius: "5px",
    background: "#95d095",
    height: "25px",
    width: "65px",
    border: "none",
    margin: "8px 13px",
    "&:hover": {
      backgroundColor: "#4CAF50",
    },
  },
  buttonClickedStyle: {
    backgroundColor: "#4CAF50",
  },
  footer: {
    textAlign: "center",
    fontSize: "14px",
    width: "100%",
    margin: "20px 0",
  },
  footerLabel: {
    fontWeight: 600,
  },
}));

export const HomePage = () => {
  const [endPoint, setEndPoint] = useState(
    "https://api.spaceXdata.com/v3/launches?limit=100"
  );
  const [selection, setSelection] = useState({
    launch_success: false,
    land_success: false,
    launch_year: 0,
  });
  const classes = useStyles();

  const yearRange = (start, end) => {
    return Array.from({ length: end - start + 1 }, (v, k) => k + start);
  };
  const filterContent = React.useCallback(() => {
    const baseEndPoint = "https://api.spaceXdata.com/v3/launches?limit=100";
    switch (true) {
      case selection.launch_success &&
        selection.land_success &&
        parseInt(selection.launch_year) !== 0:
        setEndPoint(
          baseEndPoint +
            `&launch_success=true&land_success=true&launch_year=${selection.launch_year}`
        );
        break;
      case selection.launch_success && selection.land_success:
        setEndPoint(baseEndPoint + "&launch_success=true&land_success=true");
        break;
      case selection.launch_success:
        setEndPoint(baseEndPoint + "&launch_success=true");
        break;

      default:
        return baseEndPoint;
    }
  }, [selection]);
  const handleClick = (e) => {
    const targetID = e.target.id;
    const targetValue = e.target.value;
    if (targetID === "Launch Year" && targetValue > 0) {
      setSelection({
        ...selection,
        launch_year: parseInt(targetValue),
      });
    } else if (targetID === "Successful Launch" && targetValue === "True") {
      setSelection({
        ...selection,
        launch_success: true,
      });
    } else if (targetID === "Successful Landing" && targetValue === "True") {
      setSelection({
        ...selection,
        land_success: true,
      });
    }
  };

  React.useEffect(() => {
    filterContent();
  }, [filterContent]);
  const renderYearFilter = (fieldName) => {
    const yearRangeValues = yearRange(2006, 2020);
    return yearRangeValues.map((year) => (
      <button
        className={classes.buttonStyle}
        key={year}
        value={year}
        id={fieldName}
        onClick={handleClick}
      >
        {year}
      </button>
    ));
  };
  const renderLaunchAndLandFilter = (fieldName) => {
    return CONSTANTS.BOOLEAN_BUTTONS.map((value, index) => (
      <button
        className={classes.buttonStyle}
        key={index}
        value={value}
        id={fieldName}
        onClick={handleClick}
      >
        {value}
      </button>
    ));
  };
  const renderFilter = (title, renderItems) => (
    <>
      <div className={classes.filterTitle}>{title}</div>
      <div className={classes.filterItems}>{renderItems}</div>
    </>
  );

  return (
    <>
      {" "}
      <div className={classes.mainContainer}>
        <div className={classes.header}>SpaceX Launch Programs</div>
        <div className={classes.itemContainer}>
          <div className={classes.filters}>
            Filters
            {CONSTANTS.FILTERS_CATEGORY.map((value, index) =>
              index === 0
                ? renderFilter(value, renderYearFilter(value))
                : renderFilter(value, renderLaunchAndLandFilter(value))
            )}
            {}
          </div>
          <div className={classes.launchItems}>
            <Launches endPoint={endPoint} />
          </div>
        </div>
        <div className={classes.footer}>
          <span className={classes.footerLabel}> Developed by:</span>
          {CONSTANTS.AUTHOR}
        </div>
      </div>
    </>
  );
};
