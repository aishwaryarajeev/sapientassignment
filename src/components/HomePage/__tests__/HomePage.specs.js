import React from "react";
import { shallow } from "enzyme";
import { shallowToJson } from "enzyme-to-json";
import { HomePage } from "../HomePage";

describe("<HomePage />", () => {
  let wrapper = null;
  beforeEach(() => {
    wrapper = shallow(<HomePage />);
  });
  it("should render correctly", () => {
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
