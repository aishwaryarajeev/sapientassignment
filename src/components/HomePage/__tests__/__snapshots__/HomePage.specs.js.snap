// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomePage /> should render correctly 1`] = `
<Fragment>
   
  <div
    className="makeStyles-mainContainer-1"
  >
    <div
      className="makeStyles-header-2"
    >
      SpaceX Launch Programs
    </div>
    <div
      className="makeStyles-itemContainer-3"
    >
      <div
        className="makeStyles-filters-4"
      >
        Filters
        <div
          className="makeStyles-filterTitle-6"
        >
          Launch Year
        </div>
        <div
          className="makeStyles-filterItems-7"
        >
          <button
            className="makeStyles-buttonStyle-8"
            id="Launch Year"
            key="2006"
            onClick={[Function]}
            value={2006}
          >
            2006
          </button>
          <button
            className="makeStyles-buttonStyle-8"
            id="Launch Year"
            key="2007"
            onClick={[Function]}
            value={2007}
          >
            2007
          </button>
          <button
            className="makeStyles-buttonStyle-8"
            id="Launch Year"
            key="2008"
            onClick={[Function]}
            value={2008}
          >
            2008
          </button>
          <button
            className="makeStyles-buttonStyle-8"
            id="Launch Year"
            key="2009"
            onClick={[Function]}
            value={2009}
          >
            2009
          </button>
          <button
            className="makeStyles-buttonStyle-8"
            id="Launch Year"
            key="2010"
            onClick={[Function]}
            value={2010}
          >
            2010
          </button>
          <button
            className="makeStyles-buttonStyle-8"
            id="Launch Year"
            key="2011"
            onClick={[Function]}
            value={2011}
          >
            2011
          </button>
          <button
            className="makeStyles-buttonStyle-8"
            id="Launch Year"
            key="2012"
            onClick={[Function]}
            value={2012}
          >
            2012
          </button>
          <button
            className="makeStyles-buttonStyle-8"
            id="Launch Year"
            key="2013"
            onClick={[Function]}
            value={2013}
          >
            2013
          </button>
          <button
            className="makeStyles-buttonStyle-8"
            id="Launch Year"
            key="2014"
            onClick={[Function]}
            value={2014}
          >
            2014
          </button>
          <button
            className="makeStyles-buttonStyle-8"
            id="Launch Year"
            key="2015"
            onClick={[Function]}
            value={2015}
          >
            2015
          </button>
          <button
            className="makeStyles-buttonStyle-8"
            id="Launch Year"
            key="2016"
            onClick={[Function]}
            value={2016}
          >
            2016
          </button>
          <button
            className="makeStyles-buttonStyle-8"
            id="Launch Year"
            key="2017"
            onClick={[Function]}
            value={2017}
          >
            2017
          </button>
          <button
            className="makeStyles-buttonStyle-8"
            id="Launch Year"
            key="2018"
            onClick={[Function]}
            value={2018}
          >
            2018
          </button>
          <button
            className="makeStyles-buttonStyle-8"
            id="Launch Year"
            key="2019"
            onClick={[Function]}
            value={2019}
          >
            2019
          </button>
          <button
            className="makeStyles-buttonStyle-8"
            id="Launch Year"
            key="2020"
            onClick={[Function]}
            value={2020}
          >
            2020
          </button>
        </div>
        <div
          className="makeStyles-filterTitle-6"
        >
          Successful Launch
        </div>
        <div
          className="makeStyles-filterItems-7"
        >
          <button
            className="makeStyles-buttonStyle-8"
            id="Successful Launch"
            key="0"
            onClick={[Function]}
            value="True"
          >
            True
          </button>
          <button
            className="makeStyles-buttonStyle-8"
            id="Successful Launch"
            key="1"
            onClick={[Function]}
            value="False"
          >
            False
          </button>
        </div>
        <div
          className="makeStyles-filterTitle-6"
        >
          Successful Landing
        </div>
        <div
          className="makeStyles-filterItems-7"
        >
          <button
            className="makeStyles-buttonStyle-8"
            id="Successful Landing"
            key="0"
            onClick={[Function]}
            value="True"
          >
            True
          </button>
          <button
            className="makeStyles-buttonStyle-8"
            id="Successful Landing"
            key="1"
            onClick={[Function]}
            value="False"
          >
            False
          </button>
        </div>
      </div>
      <div
        className="makeStyles-launchItems-5"
      >
        <Launches
          endPoint="https://api.spaceXdata.com/v3/launches?limit=100"
        />
      </div>
    </div>
    <div
      className="makeStyles-footer-10"
    >
      <span
        className="makeStyles-footerLabel-11"
      >
         Developed by:
      </span>
      Aishwarya Rajeev
    </div>
  </div>
</Fragment>
`;
